### ISSUE TITLE
Redis 6 Connection with Config File Specified Error

### ISSUE THREAD

**Issue by ectherrien** [StackExchange/StackExchange.Redis#1485]
Hello, I just updated to Redis 6 to obtain TLS and ACL functionalities. I have not worked on TLS yet, I figured adding a username and password would be a good place to start with the update. So currently when initializing the configurations options I added a user and password option and print out their values right before I try to connect so I know the username and password are being set properly. One thought I had was that special characters might be messing it up but not displayed in when they are printed out (had this happen to me before), but this did not fix anything so now I am here asking for assistance. Also please note redis is running locally through a docker image.

Versions:
*Redis: 6.0.4
*StackExchange.Redis: 2.1.30 (I saw there is some auth fix in 2.1.39 not sure if it applies, but I also can not test due to my build pulling nuget packages with each rebuild and 2.1.39 is not in nuget yet)

Logs/Error:
```
INFO  01:09:17,842 [1] (Root.XXX): username: admin
INFO  01:09:17,842 [1] (Root.XXX): Decrypted password: admin
FATAL 01:09:18,319 [1] (Kernel): System.ApplicationException: Error initializing component XXXXXXXXXXXXXXX
 ---> StackExchange.Redis.RedisConnectionException: It was not possible to connect to the redis server(s). UnableToConnect on redis:6379/Interactive, Initializing/NotStarted, last: NONE, origin: BeginConnectAsync, outstanding: 0, last-read: 0s ago, last-write: 0s ago, keep-alive: 60s, state: Connecting, mgr: 10 of 10 available, last-heartbeat: never, global: 0s ago, v: 2.1.30.38891
   at StackExchange.Redis.ConnectionMultiplexer.ConnectImpl(Object configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1034
   at StackExchange.Redis.ConnectionMultiplexer.Connect(ConfigurationOptions configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1002
```

Redis Logs (Show the auth credentials work):
```
127.0.0.1:6379> ping
(error) NOAUTH Authentication required.
127.0.0.1:6379> auth admin admin
OK
127.0.0.1:6379> ping
PONG
```

Code (Excludes all config other than auth, all of which is the same as it was with redis 5. FYI variable values are passed in with the following values: username = admin, password = admin)
```
ConfigurationOptions configOptions = ConfigurationOptions.Parse(hosts);

 // Adds password and/or username with all special chars removed
configOptions.User = Regex.Replace(username, @"\s+", "");
configOptions.Password = Regex.Replace(password, @"\s+", "");

mLogger.Info($"username: {configOptions.User}");
mLogger.Info($"Decrypted password: {configOptions.Password}");

mConnectionMultiplexer = ConnectionMultiplexer.Connect(configOptions);
```

Thank you!

**Comment by mgravell** [comment]
I wonder if this is misreporting a different error. My *guess* would be: .NET framework and an assembly binding redirect problem? If you enable the default user without auth: can you connect via *that*?

Testing locally, with `<PackageReference Include="StackExchange.Redis" Version="2.1.30" />` on both of `<TargetFrameworks>netcoreapp3.1;net472</TargetFrameworks>`, it works fine with a basic test here (similar to your example):

``` c#
    static void Main()
    {
        var config = ConfigurationOptions.Parse("127.0.0.1:6379");
        config.User = "admin";
        config.Password = "admin";
        using var muxer = ConnectionMultiplexer.Connect(config);
        var db = muxer.GetDatabase();
        db.KeyDelete("foo");
        for (int i = 0; i < 5; i++)
            db.StringIncrement("foo");
        Console.WriteLine(db.StringGet("foo")); // 5
    }
```

And to show redis configured for auth:

``` txt
marc@gecko:~$ redis-cli
127.0.0.1:6379> ping
(error) NOAUTH Authentication required.
127.0.0.1:6379> auth admin
(error) ERR AUTH <password> called without any password configured for the default user. Are you sure your configuration is correct?
127.0.0.1:6379> auth admin admin
OK
127.0.0.1:6379> ping
PONG
127.0.0.1:6379> acl list
1) "user admin on #8c6976e5b5410415bde908bd4dee15dfb167a9c873fc4bb8a81f6f2ab448a918 ~* +@all"
2) "user default off nopass ~* +@all"
127.0.0.1:6379> info server
# Server
redis_version:6.0.4
...
```

**Comment by ectherrien** [comment]
Thanks for the quick response! I tried with no auth and it got the same error. So then I went back to Redis 5 with no auth and it worked. I think there is some issue in my redis 6 config file. I do not think it is an issue on the StackExchange.Redis side. Closing at least till I can get Redis 6 to function properly. Thanks!

**Comment by mgravell** [comment]
If you don't figure it out, feel free to get back in touch. Let me know if you have any specific environment / setup - OS, framework, etc - the more I have to try and replicate the problem, the better. But: leaving to one side until you come back ;)

**Comment by ectherrien** [comment]
Thanks I actually just figured it out. It was more of a docker issue than anything weirdly. So under the volumes section of the docker compose yml I put a `.` in front of all of the paths and it just magically works now. Really unsure how or why that would cause an issue but it fixed my issue. If anyone else has that issue hope that helps.

Didn't work:
```
version: "3"
services:
  redis:
    image: "redis:6.0-alpine"
    ports:
      - "6379:6379"
    command: redis-server /usr/local/etc/redis/redis.conf
    volumes:
      - /opt/workspaces/XXXX/tools/docker/compose-infra/redis.conf:/usr/local/etc/redis/redis.conf
```

Works (Notice period added in front of the ` /opt/`):
```
version: "3"
services:
  redis:
    image: "redis:6.0-alpine"
    ports:
      - "6379:6379"
    command: redis-server /usr/local/etc/redis/redis.conf
    volumes:
      - ./opt/workspaces/XXXX/tools/docker/compose-infra/redis.conf:/usr/local/etc/redis/redis.conf
```

**Comment by ectherrien** [comment]
Alright I had thought I fixed it by changing the config path in the docker compose yml but what I actually did was just specify a path to a config file that does not exist. When Redis 6 does not have a config file specified I can get the client to attach and work as intended, but when Redis 6 actually gets populated with a config file I have been unable to get the client to connect. I have tried the default redis 6 config (https://raw.githubusercontent.com/antirez/redis/6.0/redis.conf) as well as my custom config. Both get the same result, and there is no auth on the default one of course. I have also tried implementing auth in the cusotm config to make sure the config file is being used and the appropriate changes are shown in the redis cli. Any ideas as to why booting redis 6 up with a config on a docker image would cause issue on the StackExchange.Redis client side?

Error (same as before):
```
FATAL 19:35:40,844 [1] (Kernel): System.ApplicationException: Error initializing component XXXXXX
 ---> StackExchange.Redis.RedisConnectionException: It was not possible to connect to the redis server(s). UnableToConnect on redis:6379/Interactive, Initializing/NotStarted, last: NONE, origin: BeginConnectAsync, outstanding: 0, last-read: 0s ago, last-write: 0s ago, keep-alive: 60s, state: Connecting, mgr: 10 of 10 available, last-heartbeat: never, global: 0s ago, v: 2.1.30.38891
   at StackExchange.Redis.ConnectionMultiplexer.ConnectImpl(Object configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1034
   at StackExchange.Redis.ConnectionMultiplexer.Connect(ConfigurationOptions configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1002
```

**Comment by ectherrien** [comment]
This is a link as to essentially how the docker image with redis is being created. 
https://medium.com/better-programming/dockerizing-and-pythonizing-redis-41b1340979de

As previously stated, I have tried the Redis 6 default config as well as my customer config and both have worked fine within the redis cli. My issue is my client with StackExchange Redis will not connect and gets the above error when a config file is sent into Redis on creation.

**Comment by ectherrien** [comment]
Updated to 2.1.50 with no changes.

**Comment by ectherrien** [comment]
Since I can get it to work with no config file I decided to use no config file and manually add admin user and reset default user in the CLI. This worked fine and I could login as intended using StackExchange.Redis and it worked. However this is not ideal, I would hope I could configure Redis with a config file for easier scalability and the ability to just copy the config file from server to server and not have to manually run CLI commands. One other thing I tried was I spun redis up with the default config and then added the users as desired manually and then did a `config rewrite` to make sure it was not an issue with how I added the users to the config file but this still gave me the same error. Any ideas or things to try? Seems incredibly weird that starting the Redis server with a config file specified makes the client unable to connect.

**Comment by ectherrien** [comment]
Alright issue is finally resolved. It did end up being a binding issue in the default redis config working with docker. Once the default bind was commented out it worked.

**Comment by kyro-dev** [comment]
> Alright issue is finally resolved. It did end up being a binding issue in the default redis config working with docker. Once the default bind was commented out it worked.

What do you mean by this? Show code please.

**Comment by mwmichael** [comment]
@kyro-dev I think @ectherrien means that within redis.conf there is this section

```################################## NETWORK #####################################

# By default, if no "bind" configuration directive is specified, Redis listens
# for connections from all available network interfaces on the host machine.
# It is possible to listen to just one or multiple selected interfaces using
# the "bind" configuration directive, followed by one or more IP addresses.
#
# Examples:
#
# bind 192.168.1.100 10.0.0.1
# bind 127.0.0.1 ::1
#
# ~~~ WARNING ~~~ If the computer running Redis is directly exposed to the
# internet, binding to all the interfaces is dangerous and will expose the
# instance to everybody on the internet. So by default we uncomment the
# following bind directive, that will force Redis to listen only on the
# IPv4 loopback interface address (this means Redis will only be able to
# accept client connections from the same host that it is running on).
#
# IF YOU ARE SURE YOU WANT YOUR INSTANCE TO LISTEN TO ALL THE INTERFACES
# JUST COMMENT OUT THE FOLLOWING LINE.
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bind 0.0.0.0 ::0```

**Comment by marksirvine** [comment]
I fixed this by commenting out the `bind 127.0.0.1 ::1` line in `/usr/local/etc/redis.conf`
